presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarFromAnnotationsParser.java

```diff
@@ -93,16 +95,32 @@ private ScalarFromAnnotationsParser() {}
 
     private static SqlScalarFunction parseParametricScalar(ScalarHeaderAndMethods scalar, Optional<Constructor<?>> constructor)
     {
-        ParametricImplementationsGroup.Builder<ScalarImplementation> implementationsBuilder = ParametricImplementationsGroup.builder();
+        ParametricImplementationsGroup.Builder<ParametricScalarImplementation> implementationsBuilder = ParametricImplementationsGroup.builder();
         ScalarImplementationHeader header = scalar.getHeader();
         checkArgument(!header.getName().isEmpty());
 
+        Map<Signature, ParametricScalarImplementation> signatures = new HashMap<Signature, ParametricScalarImplementation>();
         for (Method method : scalar.getMethods()) {
-            ScalarImplementation implementation = ScalarImplementation.Parser.parseImplementation(header.getName(), method, constructor);
-            implementationsBuilder.addImplementation(implementation);
+            ParametricScalarImplementation implementation = ParametricScalarImplementation.Parser.parseImplementation(header.getName(), method, constructor);
+            if (!signatures.containsKey(implementation.getSignature())) {
```

Comment:
Originally, we thought `Signature` itself can distinguish the different PSI. Thus we only use `Signature` as the hash key.

Now since `Signature` itself is not enough to distinguish PSI, and argument native container  + specialize native container type are required. These are issues with the current code:

- It's a bit involving
- I think the current implementation is incorrect. `signature`'s value type should be expanded to `List< ParametricScalarImplementation>` so all the PSIC with the same signature are considered. 

I think we should extract all them into a class `ParametricScalarImplKey`. Sounds reasonable ? @haozhun